File: day09/part1/main.go
// -*- compile-command: "go run main.go -p 5 ../example1.txt && go run main.go -p 25 ../input.txt"; -*-

package main

import (
	"flag"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

var (
	preamble = flag.Int("p", 25, "Preamble length")
	verbose  = flag.Bool("v", false, "Verbose log messages")
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

func process(filename string) {
	var ints []int

	lines := readLines(filename)
	logf("lines=%v", lines)
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		n, err := strconv.Atoi(line)
		check("strconv.Atoi: %v", err)
		ints = append(ints, n)
	}

	for i := *preamble; i < len(ints); i++ {
		if foundSolution(i, ints) {
			break
		}
	}
}

func foundSolution(n int, ints []int) bool {
	v := ints[n]
	seen := map[int]bool{}
	for i := n - *preamble; i < n; i++ {
		d := v - ints[i]
		if seen[d] {
			return false
		}
		seen[ints[i]] = true
	}
	log.Printf("Solution: %v", v)
	return true
}

func readLines(filename string) []string {
	log.Printf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)
	lines := strings.Split(string(buf), "\n")
	return lines
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}

File: day09/part1/main.dart
// -*- compile-command: "dart main.dart -p 5 ../example1.txt && dart main.dart -p 25 ../input.txt"; -*-

import 'dart:io';

main(List<String> args) {
  if (args.length != 3 || args[0] != '-p') {
    print('usage: dart main.dart -p 25 ../imput.txt');
    return;
  }

  var preamble = int.parse(args[1]);
  process(args[2], preamble);

  print('Done.');
}

process(String filename, int preamble) {
  print('Processing $filename ...');

  var ints = List<int>();
  var lines = File(filename).readAsLinesSync();
  for (var line in lines) {
    var n = int.parse(line);
    ints.add(n);
  }

  for (var i = preamble; i < ints.length; i++) {
    if (foundSolution(i, ints, preamble)) {
      break;
    }
  }
}

bool foundSolution(int n, List<int> ints, int preamble) {
  var v = ints[n];
  var seen = Set<int>();
  for (int i = n - preamble; i < n; i++) {
    var d = v - ints[i];
    if (seen.contains(d)) {
      return false;
    }
    seen.add(ints[i]);
  }
  print('Solution: $v');
  return true;
}
